README.md
# Fast-Reply

If like me you are spending multiple hours every day on Steem, creating 
content but also interacting with the community, you might live with the 
impression that you have too many comments to be able to answer them all… 

I love interacting with the Steem community ! But Finding all the comments 
and giving a personal answer is definitely an important part of my day. 
But it takes a lot of time… and I am always afraid to have missed one !

Fast-Reply is a web application capable of :

1. Finding all the recent (< 7 days) un-answered comments made on your 
posts and comments.
2. Providing a simple distraction-less environment to write a response
3. Going around the delay between action currently imposed by the Steem 
blockchain.

## Tool Annoncement

- 0.1: This tool has been [first annonced on Steem on May 21, 2018](https://steemit.com/utopian-io/@roxane/fast-reply-v0-1-never-miss-to-answer-a-comment-again-and-do-it-faster-than-ever).
- 0.2: First update [released on May 28, 2018](https://steemit.com/utopian-io/@roxane/fast-reply-v0-2-never-miss-to-answer-a-comment-again-and-do-it-faster-than-ever)

## Live Demo

A live demo is available: [HERE](http://fast-reply.surge.sh)

## Technical stuff

Version 0.2 is a complete re-write using a new architecture used for 
larger, team-developed projects. More precisely, the code is still written 
using the Vue.js 2 javascript framework but we now:

- Use [npm](https://www.npmjs.com/) & [Webpack](https://webpack.js.org/) to built
- Have a [components-based](https://fr.vuejs.org/v2/guide/components.html) 
source code
- Added the [vue-router](https://router.vuejs.org/) and [vuex](https://vuex.vuejs.org/) plugins. 

Those changes result in a more stable, easier to maintain and to evolve 
piece of software.

### Build Setup

``` bash
# install dependencies
npm install

# serve with hot reload at localhost:8080
npm run dev

# build for production with minification
npm run build

# build for production and view the bundle analyzer report
npm run build --report

# run unit tests
npm run unit

# run e2e tests
npm run e2e

# run all tests
npm test
```


### Security

Fast-Reply uses SteemConnect to handle authentication.
The posting role is used since the application only interact with the
Steem blockchain for basic features:

- Vote
- Comment
- Un/Follow/Mute user

## Authors

@comprendre-steem is a collective account which purpose is to 
explain Steem and help with the onboarding process of new french users.

For that purpose, a website has been created to gather all the 
important pieces of information in french. All kind of useful information 
are available on that website from markdown tutorial to Steem whitepapers
translation, but also video tutorial on Steem and its tools.

@roxane is one of the french community leaders, she has been teaching
Steem and its ecosystem for almost one year.

@oroger is more focused on the technical understanding and support
regarding the blockchain and development aspects.

## Disclaimer

This application is currently under development.  
Updates are pushed frequently and might provoke instabilities.
This is mainly due to a de-synchronisation of the localstorage.
A button will be provided to clear the localstorage and start with
a new state for those cases.

Also, the interaction with the Steem blockchain rely on the Steem API.
We noticed that sometimes, the API return error but still execute the
requested actions. Safeguards have been put in place to avoid problems
but you could still have cases where a comment or a vote has been sent multiple 
times... 

future versions will allow you to have a better view on what happens...
in the mean time any constructive feedback is welcome.

## License

The software is released under MIT License.


CHANGELOG.md
# Changelog
All notable changes to the *Fast Reply* project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1] - 2018-05-21
### Added
- Login with SteemConnect
- Filter by root post
- Voting slider
- Ability to reply/vote/vote&reply/ignore comment
- Basic emoji support

## [0.2] - 2018-05-28
### Added
- Architecture improvements
- UI improvements
  - Tooltips
  - Busy links
  - Highlight selected comment
  - Sort comments by date ASC/DESC
  - Negative votes (flag)
- Editor imrovements
  - Image drag & drop
  - Side by side Markdown preview
- Pending action panel
  - Pause / Resume
  - Retry 3 times before giving up
  - Force execute / delete pending action
- Mentions panel

## [0.3] - 2018-06-??
### Added
- Hotkeys
- Sentry support
- News menu item
- New Logo

### Changed
- Signature is now customizable
- Support multiple config, one per username


test/unit/jest.conf.js
const path = require('path')

module.exports = {
  rootDir: path.resolve(__dirname, '../../'),
  moduleFileExtensions: [
    'js',
    'json',
    'vue'
  ],
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1'
  },
  transform: {
    '^.+\\.js$': '<rootDir>/node_modules/babel-jest',
    '.*\\.(vue)$': '<rootDir>/node_modules/vue-jest'
  },
  testPathIgnorePatterns: [
    '<rootDir>/test/e2e'
  ],
  snapshotSerializers: ['<rootDir>/node_modules/jest-serializer-vue'],
  setupFiles: ['<rootDir>/test/unit/setup'],
  //mapCoverage: true,
  coverageDirectory: '<rootDir>/test/unit/coverage',
  collectCoverageFrom: [
    'src/**/*.{js,vue}',
    '!src/main.js',
    '!src/router/index.js',
    '!**/node_modules/**'
  ]
}


src/main.js
// The Vue build version to load with the `import` command
// (runtime-only or standalone) has been set in webpack.base.conf with an alias.
import Vue from 'vue'
import Raven from 'raven-js'
import RavenVue from 'raven-js/plugins/vue'

// Load 3rd patry libs
import 'bulma/css/bulma.css'
import 'bulma-slider/dist/bulma-slider.min.css'
import 'bulma-slider/dist/bulma-slider.min'
import 'vue-awesome/icons'
import Icon from 'vue-awesome/components/Icon'

// Initialize App
import './utils/filters.js'
import store from '@/utils/store.js'
import router from '@/utils/router'
import App from '@/App'

// Global component declaration
Vue.component('icon', Icon)

Vue.config.productionTip = false

// Sentry.io configuration
Raven
  .config('https://b8ee76cd3ced4d5aa02d88cae485d9fa@sentry.io/1215865', {release: '0.3-beta'})
  .addPlugin(RavenVue, Vue)
  .install()

/* eslint-disable no-new */
new Vue({
  el: '#app',
  store,
  router,
  components: { App },
  template: '<App/>',
  created () {
    console.log('Load Steem profile')
    // Request user details if token is available
    let token = this.$ls.get('access_token')
    if (token != null) {
      this.$store.dispatch('connect', token)
    }
  },
  mounted () {
    const SECOND = 1000

    /** Start infinite execution of trigger action **/
    // https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setInterval

    // refresh VP every minute
    let vpTimer = setInterval(this.$store.dispatch, 60 * SECOND, 'updateVP')
    this.$store.dispatch('timer', {name: 'updateVP', value: vpTimer})

    // execute a pending vote every 5 seconds
    let voteTimer = setInterval(this.$store.dispatch, 5 * SECOND, 'executeNextPendingActionOfType', 'vote')
    this.$store.dispatch('timer', {name: 'executeNextPendingVote', value: voteTimer})

    // execute a pending comment every 30 seconds
    let commentTimer = setInterval(this.$store.dispatch, 30 * SECOND, 'executeNextPendingActionOfType', 'comment')
    this.$store.dispatch('timer', {name: 'executeNextPendingComment', value: commentTimer})
  }
})


src/utils/filters.js
// https://stackoverflow.com/a/45054257/957103

import Vue from 'vue'
import showdown from 'showdown'

Vue.filter('truncate', function (value, length) {
  // Remove words making a string larger than a given size
  if (!value) return ''
  if (value.length <= length) return value

  // https://stackoverflow.com/a/33379772/957103
  function truncateOnWord (str, limit) {
    var trimmable = '\u0009\u000A\u000B\u000C\u000D\u0020\u00A0\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u2028\u2029\u3000\uFEFF'
    var reg = new RegExp('(?=[' + trimmable + '])')
    var words = str.split(reg)
    var count = 0
    return words.filter(function (word) {
      count += word.length
      return count <= limit
    }).join('')
  }

  return truncateOnWord(value, length) + ' [...]'
})

Vue.filter('markdownToHTML', function (value) {
  // TODO: check if content is indeed markdown (via API)
  // Transform Markdown markup into HTML
  // ShowDownJS: https://github.com/showdownjs/showdown
  return new showdown.Converter({
    tables: true,
    tablesHeaderId: true,
    ghMentions: true,
    ghMentionsLink: 'https://steemit.com/@{u}',
    strikethrough: true,
    simplifiedAutoLink: true,
    excludeTrailingPunctuationFromURLs: true,
    smoothLivePreview: true,
    disableForced4SpacesIndentedSublists: true,
    simpleLineBreaks: true
  }).makeHtml(value)
})

Vue.filter('date', function (date) {
  // Format date from UTC to locale Date
  return new Date(Date.parse(date)).toLocaleDateString()
})

Vue.filter('timestamp', function (timestamp) {
  // Format timestamp from UTC to locale Date
  let ts = new Date(timestamp)
  return ts.toLocaleDateString() + ' @ ' + ts.toLocaleTimeString()
})

Vue.filter('steemit', function (path) {
  return 'https://www.steemit.com' + path
})

Vue.filter('busy', function (path) {
  return 'https://www.busy.org' + path
})

Vue.filter('profile', function (username, platform) {
  // Creation d'un lien vers le profile d'un utilisateur
  switch (platform) {
    case 'busy' :
      return 'https://www.busy.org/@' + username
    default :
      return 'https://www.steemit.com/@' + username
  }
})

Vue.filter('avatar', function (username) {
  // Creation d'un lien vers l'avatar (image) steem d'un utilisateur // Thanks to Busy
  return 'http://img.busy.org/@' + username
})


src/utils/store.js
// https://stackoverflow.com/questions/40564071/how-do-i-break-up-my-vuex-file
import Vue from 'vue'
import Vuex from 'vuex'
import VueResource from 'vue-resource' // https://github.com/pagekit/vue-resource
import VueStorage from 'vue-ls'
// Steem Libs
import steem from 'steem'
import sc2 from 'sc2-sdk'
import sc2Utils from '@/utils/sc2-utils.js'
// Utils
import Raven from 'raven-js'
import toast from '@/utils/toast.js'

// Plugin declaration
Vue.use(Vuex)
Vue.use(VueResource)
Vue.use(VueStorage)

// Initialize SteemConnect v2
var api = sc2.Initialize({
  app: 'fast-reply.app',
  callbackURL: 'http://localhost:8080/#/steemconnect', // Dev localhost URL
  // callbackURL: 'http://very-fast-reply.surge.sh/#/steemconnect', // Beta URL
  // callbackURL: 'http://fast-reply.surge.sh/#/steemconnect/', // Live demo URL
  scope: ['vote', 'comment', 'custom_json']
})

let defaultSort = {
  field: 'created',
  inverted: false
}

const defaultSignature = '<div class="pull-right"><sub><a href="https://steemit.com/@fast-reply">Sent with Fast-Reply</a></sub></div>'

let defaultConfig = {
  version: 0.3,
  ignoreList: {comments: [], users: []},
  vote: 100,
  // Since 0.2
  sort: defaultSort,
  // Since 0.3
  signature: defaultSignature
}

const filterIgnored = function (comments, filter, ignoreList) {
  return comments ? comments
    // Filter using selected article
    .filter(comment => (filter == null || filter.id === comment.rootId))
    // Filter comments from ignored users && Filter comments previously ignored
    .filter(comment => (!ignoreList.users.includes(comment.author) && !ignoreList.comments.includes(comment.id)))
    // Otherwise return null
    : null
}

const LS_CONFIG = 'config'
const LS_PENDING = 'pending'
const LS_RUNNING = 'running'

// Create Global Store
export default new Vuex.Store({
  state: {
    steemconnect: {
      api: api,
      user: null,
      vp: null
    },
    config: defaultConfig,
    timers: {},
    inbox: {
      comments: null,
      filter: null,
      selectedComment: null
    },
    pending: Vue.ls.get(LS_PENDING, []),
    isSchedulerRunning: Vue.ls.get(LS_RUNNING, true),
    accounts: {
      blog: new Set(),
      ignore: new Set()
    }
  },
  mutations: {
    clearIgnoreList (state) {
      state.config.ignoreList = {comments: [], users: []}
      this.commit('saveConfig')
    },
    connect (state, result) {
      state.steemconnect.user = result.account
    },
    updateVP (state, vp) {
      state.steemconnect.vp = vp
    },
    logout (state) {
      // clear steemconnect state
      state.steemconnect.user = null
      state.steemconnect.vp = null
    },
    loadConfig (state) {
      console.log('Loading config...')
      // First, attempt to load 0.2 config
      state.config = Vue.ls.get(LS_CONFIG, defaultConfig)
      // Then overload with 0.3 config, if present
      state.config = Vue.ls.get(LS_CONFIG + '_' + state.steemconnect.user.name, state.config)
      // Config migration 0.1 => 0.2
      if (!state.config.version) {
        state.config['sort'] = {field: 'created', inverted: false}
        state.config['version'] = 0.2
        console.log('Configuration migrated to 0.2')
      }
      // Config migration 0.2 => 0.3
      if (state.config.version === 0.2) {
        state.config['signature'] = defaultSignature
        state.config['version'] = 0.3
        console.log('Configuration migrated to 0.3')
      }
      this.commit('saveConfig')
    },
    config (state, {name, value}) {
      state.config[name] = value
      this.commit('saveConfig')
    },
    saveConfig (state) {
      const key = LS_CONFIG + '_' + state.steemconnect.user.name
      Vue.ls.set(key, state.config)
    },
    timer (state, timer) {
      if (state.timers[timer]) {
        // Remove previous interval, if set
        clearInterval(state.timers[timer.name])
        state.timers[timer] = null
      }

      if (timer.value) {
        state.timers[timer.name] = timer.value
      }
    },
    reload (state, result) {
      state.inbox.comments = result.data.comments
      // Count number of unique rootId among comments
      let articlesCount = state.inbox.comments.map(c => c.rootId).filter((value, index, self) => self.indexOf(value) === index).length
      toast.createDialog('success', 'Found ' + state.inbox.comments.length + ' new comment(s) on ' + articlesCount + ' articles', 5000)
    },
    selectFilter (state, article) {
      state.inbox.filter = article
    },
    loadMessages (state, messages) {
      state.inbox.messages = messages
    },
    selectComment (state, comment) {
      state.inbox.selectedComment = comment
    },
    ignoreComment (state, commentId) {
      // Find comment based on its id and remove it from the list
      for (var i = 0; i < state.inbox.comments.length; i++) {
        if (state.inbox.comments[i].id === commentId) {
          // Find the right index and remove it
          state.inbox.comments.splice(i, 1)
        }
      }
      // Push to ignore list
      state.config.ignoreList.comments.push(commentId)
      this.commit('saveConfig')
    },

    /** Managing pending actions **/
    addPendingAction (state, action) {
      state.pending.push(action)
      Vue.ls.set(LS_PENDING, state.pending)
    },
    deletePendingAction (state, action) {
      state.pending = state.pending.filter(a => a !== action)
      Vue.ls.set(LS_PENDING, state.pending)
    },
    failedAttempt (state, action) {
      // Search for action in state
      let failedAction = state.pending.find(item => item === action)
      if (failedAction) {
        // if found, increase attempts counter
        failedAction.attempts++
      }
    },
    isSchedulerRunning (state, value) {
      state.isSchedulerRunning = value
      Vue.ls.set(LS_RUNNING, state.isSchedulerRunning)
    },
    accounts (state, {type, accounts}) {
      state.accounts[type] = accounts
      if (type === 'ignore') {
        // Use those names as ignore list to filter the messages
        state.config.ignoreList.users = Array.from(accounts)
        this.commit('saveConfig')
      }
    }
  },
  actions: {
    async connect ({ dispatch, commit, state }, token) {
      let success = false
      // wait for async call to resolve using await, then use result
      state.steemconnect.api.setAccessToken(token)
      await state.steemconnect.api.me()
        .then(result => {
          commit('connect', result)
          success = true
          return result
        })
        .catch(err => {
          console.log(err)
          Raven.captureException(err)
        })

      if (success) {
        // Load connected user configuration, if present
        commit('loadConfig')
        // Force reload on login
        dispatch('reload')
        // Retrieve user VP
        dispatch('updateVP')
        // Load account (un)follow/mute list
        dispatch('loadAccounts')
      } else {
        toast.createDialog('error', 'Error during login', 5000)
      }
    },
    updateVP ({ dispatch, commit, state }) {
      if (state.steemconnect.user) {
        steem.api.getAccounts([state.steemconnect.user.name], function (err, response) {
          if (!err) {
            let secondsago = (new Date() - new Date(response[0].last_vote_time + 'Z')) / 1000
            let vpow = response[0].voting_power + (10000 * secondsago / 432000)
            vpow = Math.min(vpow / 100, 100).toFixed(2)
            commit('updateVP', vpow)
          }
        })
      }
    },
    loadAccounts ({dispatch, commit, state}) {
      if (state.steemconnect.user) {
        const chunkSize = 100
        // Recursive function to go through all the following by chunk
        let loadFollowing = function (username, type, pos, accounts) {
          steem.api.getFollowing(username, pos, type, chunkSize, function (err, result) {
            if (!err) {
              for (let follows of result) {
                if (follows.what.length > 0 && follows.what[0] === type) {
                  accounts.add(follows.following)
                }
              }
              // Check if there is more to process
              if (result.length === chunkSize) {
                loadFollowing(username, type, result[result.length - 1].following, accounts)
              } else {
                commit('accounts', {type: type, accounts: accounts})
              }
            } else {
              console.log(err)
              toast.createDialog('error', 'Could not load related accounts: ' + err, 2000)
              Raven.captureException(err)
            }
          })
        }
        // Start fetching the account data
        loadFollowing(state.steemconnect.user.name, 'blog', 0, new Set())
        loadFollowing(state.steemconnect.user.name, 'ignore', 0, new Set())
      }
    },
    clearIgnoreList ({ dispatch, commit }) {
      commit('clearIgnoreList')
      dispatch('reload')
    },
    logout ({ commit }) {
      commit('logout')
    },
    timer ({dispatch, commit, state}, timer) {
      commit('timer', timer)
    },
    async reload ({dispatch, commit, state}) {
      let user = state.steemconnect.user
      if (user) {
        const url = 'http://api.comprendre-steem.fr/getComments?username=' + user.name
        commit('reload', await Vue.http.get(url))
        dispatch('selectFirstComment')
      }
    },
    selectFirstComment ({dispatch, commit, state}) {
      // Filter to find first non ignored comment
      let filtered = filterIgnored(state.inbox.comments, state.inbox.filter, state.config.ignoreList)
      if (filtered.length > 0) {
        let position = state.config.sort.inverted ? 0 : filtered.length - 1
        // Select that comment
        commit('selectComment', filtered[position])
      }
    },
    selectFilter ({dispatch, commit}, article) {
      commit('selectFilter', article)
    },
    selectComment ({dispatch, commit, state}, comment) {
      commit('selectComment', comment)
    },
    config ({dispatch, commit, state}, {key, value}) {
      commit('config', {name: key, value: value})
    },
    markCommentProcessed ({dispatch, commit, state}, commentId) {
      commit('ignoreComment', commentId)
      dispatch('selectFirstComment')
    },
    addUserToIgnore ({dispatch, commit, state}, username) {
      // Blockchain has been updated, let's reload those data locally too
      commit('loadAccounts')
    },

    /** Managing pending actions **/
    addPendingAction ({dispatch, commit, state}, action) {
      commit('addPendingAction', action)
    },
    deletePendingAction ({dispatch, commit, state}, action) {
      commit('deletePendingAction', action)
    },
    executePendingAction ({dispatch, commit, state}, action) {
      switch (action.type) {
        case 'comment': {
          sc2Utils.comment(state.steemconnect.api, state.steemconnect.user.name, action.author, action.permlink, action.body, action.created)
            .then(() => commit('deletePendingAction', action))
            .catch(err => {
              console.log('Error SC2 during comment: ', err.error_description)
              console.log('Failed action: ', action)
              commit('failedAttempt', action)
              Raven.captureException(err)
            })
          break
        }
        case 'vote': {
          sc2Utils.vote(state.steemconnect.api, state.steemconnect.user.name, action.author, action.permlink, action.vote)
            .then(() => commit('deletePendingAction', action))
            .catch(err => {
              commit('failedAttempt', action)
              if (err.error_description === 'itr->vote_percent != o.weight: You have already voted in a similar way.') {
                // already been processed, remove this action from pending
                commit('deletePendingAction', action)
              } else {
                console.log('Failed action: ', action)
                console.log('Error SC2 during vote: ', err.error_description)
                commit('failedAttempt', action)
                Raven.captureException(err)
              }
            })
          break
        }
        default: {
          toast.createDialog('error', 'Unknown action: ' + action.type, 3000)
          break
        }
      }
    },
    /** Execute next action if scheduler is enabled **/
    executeNextPendingActionOfType ({dispatch, commit, state}, type) {
      if (state.isSchedulerRunning) {
        // find next pending action of requested type, if any
        let next = state.pending
          .filter(action => action.type === type)
          // Abandon the action after 5 failed attempts
          .filter(action => action.attempts < 3)
          // If candidates are available, take the first one
          .shift()
        if (next) {
          // Execute it
          dispatch('executePendingAction', next)
        }
      }
    },
    toggleScheduler ({commit, state}) {
      commit('isSchedulerRunning', !state.isSchedulerRunning)
    }
  },
  getters: {
    user: state => {
      return state.steemconnect.user
    },
    steemconnect: state => {
      return state.steemconnect
    },
    config: state => {
      return state.config
    },
    inbox: state => {
      return {
        filter: state.inbox.filter,
        selectedComment: state.inbox.selectedComment,
        comments: filterIgnored(state.inbox.comments, null, state.config.ignoreList)
      }
    },
    pending: state => {
      return state.pending
    },
    accounts: state => {
      return state.accounts
    },
    isSchedulerRunning: state => {
      return state.isSchedulerRunning
    }
  }
})
